### ISSUE TITLE
Running RAP with LLMs Missing `get_next_token_logits` and `get_loglikelihood`

### ISSUE THREAD

**Issue by ERoszczyk** [maitrix-org/llm-reasoners#153]
I've noticed that many LLMs available through APIs do not support functions like `get_next_token_logits` or `get_loglikelihood`. Since RAP relies on these functions, I'm wondering if it's possible to run RAP with an LLM that doesn't provide them.

I'm particularly interested in the prontoqa implementation using RAP. However, I found that both of these functions are used in the `fast_reward` function within `ProntoQAConfig` (see [this link](https://github.com/maitrix-org/llm-reasoners/blob/259fa0c8c62ba3dc9dff2883f165b06c88639bea/examples/RAP/prontoqa/search_config.py#L57)). Although there’s a comment referring to this as "OLD fast reward code," it appears that this implementation is still in use across several parts of the codebase.

Could someone clarify whether there’s an alternative approach for LLMs that lack these functions, or if this part of the implementation is planned for replacement? Any insights or suggestions would be appreciated.

**Comment by Ber666** [comment]
Thanks for your question! We do plan to update the code to remove the reliance on logits and likelihood, but we're not sure when it will be ready, as we're currently short on manpower. Let me know if you're interested in contributing!

**Comment by pangpang-xuan** [comment]
```
def fast_reward(self, state: MedqaState, action: MedqaAction) -> tuple[float, dict]:
      with io.StringIO() as f:
                  f.write(self.useful_prompt["input"])
                  f.write(self.useful_prompt["question_prefix"] + self.example + "\n")
                  for idx, (q, _, _) in enumerate(state):
                      f.write(self.useful_prompt["subquestion_prefix"].format(idx + 1) + " " + q + "\n")
                  f.write(self.useful_prompt["new_subquestion_prefix"].format(len(state) + 1) + " " + action + "\n")
                  f.write(self.useful_prompt["useful_prefix"])
                  model_input = f.getvalue()
      if type(self.base_model) is reasoners.lm.hf_model.HFModel:
                  print("type is HFModel class")
                  logits = self.base_model.get_next_token_logits(model_input, ["Yes", "No"])[0]
                  probs = np.exp(logits) / np.sum(np.exp(logits))
                  useful_prob = probs[0]
                  print("--------------- fast_reward result is ----------------------")
                  print(useful_prob)
                  print("--------------- fast_reward result over ---------------------")
                  fast_reward, _ = self.calculate_reward(useful_prob)
                  return fast_reward, {'r_useful': useful_prob}
              
              elif type(self.base_model) is reasoners.lm.vllm_model.VLLMModel or type(self.base_model) is reasoners.lm.openai_model.OpenAIModel or type(self.base_model) is reasoners.lm.deepseek_model.DeepseekModel:
                  print("type is vllm/openai/deepseek class")
                  probs = self.base_model.get_next_token_logits(model_input,["Yes", "No"])
                  useful_prob = probs[0]
                  print("--------------- fast_reward result is ----------------------")
                  print(useful_prob)
                  print("--------------- fast_reward result over ---------------------")
                  fast_reward, _ = self.calculate_reward(useful_prob)
                  return fast_reward, {'r_useful': useful_prob}
              else:
                  raise Exception('Model Type Error!')

def get_next_token_logits(
        self,
        prompt: Union[str, list[str]],
        candidates: Union[list[str], list[list[str]]],
        **kwargs,
    ) -> list[np.ndarray]:
        
        import numpy as np

        score = 0

        for _ in range(5):
            try:
                response = self.client.chat.completions.create( 
                    model= self.model,
                    messages=[
                        {"role": "system", "content": "You are a helpful assistant and you can only answer Yes or No"},
                        {"role": "user", "content": prompt},
                    ],
                    stream=False
                )
                content = str(response.choices[0].message.content)
                if "Yes" in content or "yes" in content:
                    score += 1
            except Exception as e:
                print(f"An error occurred: {e}")

        yes_prob, no_prob = score / 5, (5 - score) / 5
        probs = np.array([yes_prob, no_prob]) 

        return probs

```

i use this method to solve this problem but i dont know whether its right. i hope this can give u an idea to solve this problem and hope the llm-reasoners creators can give me a advice to my solution.
